app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -0,0 +1,428 @@
+     * @return the future that indicate if the username is available
+     */
+    fun isUsernameAvailable(userName: String): CompletableFuture<Boolean> {
+        return database.isUsernameAvailable(userName)
+    }
+
+    /**
+     * Use this function to modify the username(the username will be modify only if it is available on the database)
+     * @param username that we want to set
+     */
+    fun setUsername(username: String): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setUsername(username).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateUsername(currentUserID!!, username)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily distance goal of the user
+     * @param distanceGoal new daily distance goal
+     */
+    fun setDistanceGoal(distanceGoal: Double): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkDistanceGoal(distanceGoal)
+        return CompletableFuture.supplyAsync {
+            database.setDistanceGoal(distanceGoal).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateDistanceGoal(currentUserID!!, distanceGoal)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily activity time goal of the user
+     * @param activityTimeGoal new daily activity time goal
+     */
+    fun setActivityTimeGoal(activityTimeGoal: Double): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkActivityTimeGoal(activityTimeGoal)
+        return CompletableFuture.supplyAsync {
+            database.setActivityTimeGoal(activityTimeGoal).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updateTimeGoal(currentUserID!!, activityTimeGoal)
+            }
+            it
+        }
+    }
+
+    /**
+     * Use this function to modify the daily number of paths goal of the user
+     * @param nbOfPathsGoal new daily number of paths goal
+     */
+    fun setNumberOfPathsGoal(nbOfPathsGoal: Int): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        checkNbOfPathsGoal(nbOfPathsGoal)
+        return CompletableFuture.supplyAsync {
+            database.setNbOfPathsGoal(nbOfPathsGoal).join()
+        }.thenApplyAsync {
+            if (it) {
+                cache.updatePathsGoal(currentUserID!!, nbOfPathsGoal)
+            }
+            it
+        }
+    }
+
+    /**
+     * This function will set the photo as the profile photo of the user
+     * @param photo that we want to set
+     * @return a completable future that indicate if the photo was correctly stored
+     */
+    fun setProfilePhoto(photo: Bitmap): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setProfilePhoto(photo).join()
+        }.thenApplyAsync {
+            if (it) {

```

Comment:
Same remark here : I think it is incorrect , you should write instead : return database.setProfilePhoto(photo).thenApply{ isSet ->
if(isSet) cache.updatePhoto(currentUserID!!, UserEntity.fromBitmapToByteArray(photo))
isSet
}